# **Waze Project**
**Course 2 - Get Started with Python**

Welcome to the Waze Project!

Your Waze data analytics team is still in the early stages of their user churn project. Previously, you were asked to complete a project proposal by your supervisor, May Santner. You have received notice that your project proposal has been approved and that your team has been given access to Waze's user data. To get clear insights, the user data must be inspected and prepared for the upcoming process of exploratory data analysis (EDA).

A Python notebook has been prepared to guide you through this project. Answer the questions and create an executive summary for the Waze data team.

# **Course 2 End-of-course project: Inspect and analyze data**

In this activity, you will examine data provided and prepare it for analysis. This activity will help ensure the information is,

1.   Ready to answer questions and yield insights

2.   Ready for visualizations

3.   Ready for future hypothesis testing and statistical methods
<br/>

**The purpose** of this project is to investigate and understand the data provided.

**The goal** is to use a dataframe contructed within Python, perform a cursory inspection of the provided dataset, and inform team members of your findings.
<br/>

*This activity has three parts:*

**Part 1:** Understand the situation
* How can you best prepare to understand and organize the provided information?

**Part 2:** Understand the data

* Create a pandas dataframe for data learning, future exploratory data analysis (EDA), and statistical activities

* Compile summary information about the data to inform next steps

**Part 3:** Understand the variables

* Use insights from your examination of the summary data to guide deeper investigation into variables


<br/>

Follow the instructions and answer the following questions to complete the activity. Then, you will complete an Executive Summary using the questions listed on the PACE Strategy Document.

Be sure to complete this activity before moving on. The next course item will provide you with a completed exemplar to compare to your own work.



# **Identify data types and compile summary information**


<img src="images/Pace.png" width="100" height="100" align=left>

# **PACE stages**

Throughout these project notebooks, you'll see references to the problem-solving framework, PACE. The following notebook components are labeled with the respective PACE stages: Plan, Analyze, Construct, and Execute.

<img src="images/Plan.png" width="100" height="100" align=left>


## **PACE: Plan**

Consider the questions in your PACE Strategy Document and those below to craft your response:

### **Task 1. Understand the situation**

*   How can you best prepare to understand and organize the provided driver data?


*Begin by exploring your dataset and consider reviewing the Data Dictionary.*

Always have the project task in mind. Start by looking at the file name(s) to roughly determine what kind of information might be contained in it. Then, open each file and browse through all titles, column names and row names to get further overview of the dataset. In all of this, ask yourself if the data is relevant to solving the project task.

<img src="images/Analyze.png" width="100" height="100" align=left>

## **PACE: Analyze**

Consider the questions in your PACE Strategy Document to reflect on the Analyze stage.

### **Task 2a. Imports and data loading**

Start by importing the packages that you will need to load and explore the dataset. Make sure to use the following import statements:

*   `import pandas as pd`

*   `import numpy as np`


In [1]:
# Import packages for data manipulation

import numpy as np
import pandas as pd


Then, load the dataset into a dataframe. Creating a dataframe will help you conduct data manipulation, exploratory data analysis (EDA), and statistical activities.

**Note:** As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [2]:
# Load dataset into dataframe
df = pd.read_csv('/Users/aronrana/Documents/Google Coursera/Google Advanced Data Analytics/Course_2/waze_dataset.csv')

### **Task 2b. Summary information**

View and inspect summary information about the dataframe by **coding the following:**

1.   df.head(10)
2.   df.info()

*Consider the following questions:*

1. When reviewing the `df.head()` output, are there any variables that have missing values?

2. When reviewing the `df.info()` output, what are the data types? How many rows and columns do you have?

3. Does the dataset have any missing values?

In [3]:
df.head(10)


ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   0  retained       283     226      296.748273                     2276   
1   1  retained       133     107      326.896596                     1225   
2   2  retained       114      95      135.522926                     2651   
3   3  retained        49      40       67.589221                       15   
4   4  retained        84      68      168.247020                     1562   
5   5  retained       113     103      279.544437                     2637   
6   6  retained         3       2      236.725314                      360   
7   7  retained        39      35      176.072845                     2999   
8   8  retained        57      46      183.532018                      424   
9   9   churned        84      68      244.802115                     2997   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     208                       0       2628.845068   
1                      19                      64      13715.920550   
2                       0                       0       3059.148818   
3                     322                       7        913.591123   
4                     166                       5       3950.202008   
5                       0                       0        901.238699   
6                     185                      18       5249.172828   
7                       0                       0       7892.052468   
8                       0                      26       2651.709764   
9                      72                       0       6043.460295   

   duration_minutes_drives  activity_days  driving_days   device  
0              1985.775061             28            19  Android  
1              3160.472914             13            11   iPhone  
2              1610.735904             14             8  Android  
3               587.196542              7             3   iPhone  
4              1219.555924             27            18  Android  
5               439.101397             15            11   iPhone  
6               726.577205             28            23   iPhone  
7              2466.981741             22            20   iPhone  
8              1594.342984             25            20  Android  
9              2341.838528              7             3   iPhone

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


1. No.
2. Data types: float64(3), int64(8), object(2). Number of columns: 13, number of rows: 14,999.
3. Yes, under the 'label' column.

### **Task 2c. Null values and summary statistics**

Compare the summary statistics of the 700 rows that are missing labels with summary statistics of the rows that are not missing any values.

**Question:** Is there a discernible difference between the two populations?


In [5]:
# Isolate rows with null values
mask_1 = (df['label'] != 'retained') & (df['label'] != 'churned')
null_values = df[mask_1]

# Display summary stats of rows with null values
null_values.describe()


ID    sessions      drives  total_sessions  \
count    700.000000  700.000000  700.000000      700.000000   
mean    7405.584286   80.837143   67.798571      198.483348   
std     4306.900234   79.987440   65.271926      140.561715   
min       77.000000    0.000000    0.000000        5.582648   
25%     3744.500000   23.000000   20.000000       94.056340   
50%     7443.000000   56.000000   47.500000      177.255925   
75%    11007.000000  112.250000   94.000000      266.058022   
max    14993.000000  556.000000  445.000000     1076.879741   

       n_days_after_onboarding  total_navigations_fav1  \
count               700.000000              700.000000   
mean               1709.295714              118.717143   
std                1005.306562              156.308140   
min                  16.000000                0.000000   
25%                 869.000000                4.000000   
50%                1650.500000               62.500000   
75%                2508.750000              169.250000   
max                3498.000000             1096.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count              700.000000        700.000000               700.000000   
mean                30.371429       3935.967029              1795.123358   
std                 46.306984       2443.107121              1419.242246   
min                  0.000000        290.119811                66.588493   
25%                  0.000000       2119.344818               779.009271   
50%                 10.000000       3421.156721              1414.966279   
75%                 43.000000       5166.097373              2443.955404   
max                352.000000      15135.391280              9746.253023   

       activity_days  driving_days  
count     700.000000    700.000000  
mean       15.382857     12.125714  
std         8.772714      7.626373  
min         0.000000      0.000000  
25%         8.000000      6.000000  
50%        15.000000     12.000000  
75%        23.000000     18.000000  
max        31.000000     30.000000

In [6]:
null_values.head(30)

ID label  sessions  drives  total_sessions  n_days_after_onboarding  \
77    77   NaN        63      50      133.104155                      783   
80    80   NaN       116      93      436.060183                     1584   
98    98   NaN        78      64      583.492789                     3414   
111  111   NaN       106     102      113.379056                     2228   
142  142   NaN        32      26      222.129310                      208   
162  162   NaN         3       3       17.250220                     3203   
176  176   NaN       283     226      529.533944                     1110   
199  199   NaN        27      23      196.839605                     2800   
266  266   NaN        70      58      137.151279                     3264   
283  283   NaN       170     137      354.297662                      869   
288  288   NaN       199     159      248.560854                     2959   
301  301   NaN        73      60      268.328520                      164   
309  309   NaN        95      81      140.093974                       58   
313  313   NaN       207     175      351.947218                      422   
330  330   NaN       121      97      203.603451                      405   
333  333   NaN        87      70      253.140234                     2461   
446  446   NaN        33      30       43.433247                     1240   
454  454   NaN         2       2       46.741886                     1680   
485  485   NaN         0       0       90.747466                      579   
495  495   NaN         7       6       62.101853                      202   
574  574   NaN       116      96      124.967244                      952   
592  592   NaN        38      31      206.557805                     1852   
658  658   NaN         6       6      200.854687                      692   
677  677   NaN        16      16      237.706155                     1598   
690  690   NaN         4       4       64.271061                     3375   
712  712   NaN        86      71      261.824509                      106   
718  718   NaN       400     321      497.118054                     1884   
747  747   NaN       268     215      302.281837                     1185   
753  753   NaN        18      14      347.135984                     1728   
773  773   NaN       140     125      345.064890                     3458   

     total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
77                      201                       0       2649.015822   
80                      283                      62       4183.409514   
98                        0                       0       1811.140893   
111                      14                       0       2817.481840   
142                      55                      10       2459.816477   
162                      22                       0      11819.366330   
176                       6                       0       3028.601300   
199                     156                       0      11579.565100   
266                      45                       0       1305.946778   
283                      18                      25       3914.401147   
288                     127                       0       5285.296516   
301                      12                      97       1811.870085   
309                     151                      79       5847.252244   
313                     119                      17       2090.096785   
330                      16                       0       1572.853676   
333                     230                       0       3654.784978   
446                       0                       0       1608.596461   
454                      51                       0       6822.713755   
485                     229                       0       1363.206140   
495                      57                     148       5245.298660   
574                      34                       0       4626.352023   
592         

In [7]:
# Isolate rows without null values
mask_2 = (df['label'] == 'retained') | (df['label'] == 'churned')
not_null_values = df[mask_2]

# Display summary stats of rows without null values
not_null_values.describe()

ID      sessions        drives  total_sessions  \
count  14299.000000  14299.000000  14299.000000    14299.000000   
mean    7503.573117     80.623820     67.255822      189.547409   
std     4331.207621     80.736502     65.947295      136.189764   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.457733   
50%     7504.000000     56.000000     48.000000      158.718571   
75%    11257.500000    111.000000     93.000000      253.540450   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14299.000000            14299.000000   
mean               1751.822505              121.747395   
std                1008.663834              147.713428   
min                   4.000000                0.000000   
25%                 878.500000               10.000000   
50%                1749.000000               71.000000   
75%                2627.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14299.000000      14299.000000             14299.000000   
mean                29.638296       4044.401535              1864.199794   
std                 45.350890       2504.977970              1448.005047   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2217.319909               840.181344   
50%                  9.000000       3496.545617              1479.394387   
75%                 43.000000       5299.972162              2466.928876   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14299.000000  14299.000000  
mean       15.544653     12.182530  
std         9.016088      7.833835  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000

Looking at average values for variables in both populations, there is not significant difference. There are differences in minimum and maximum values for certain variable in both populations. But, considering the average values, which gives a better picture of the whole pupulation, there is no major difference between the two populations. 


### **Task 2d. Null values - device counts**

Next, check the two populations with respect to the `device` variable.

**Question:** How many iPhone users had null values and how many Android users had null values?

In [8]:
# Get count of null values by device
null_values['device'].value_counts()

device
iPhone     447
Android    253
Name: count, dtype: int64

iPhone     447
Android    253

Now, of the rows with null values, calculate the percentage with each device&mdash;Android and iPhone. You can do this directly with the [`value_counts()`](https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html) function.

In [9]:
# Calculate % of iPhone nulls and Android nulls
pct_null_iphone = ((null_values['device'].value_counts()['iPhone'])/len(null_values))*100
pct_null_android = ((null_values['device'].value_counts()['Android'])/len(null_values))*100


print(pct_null_iphone)
print(pct_null_android)


63.857142857142854
36.142857142857146


How does this compare to the device ratio in the full dataset?

In [10]:
# Calculate % of iPhone users and Android users in full dataset
pct_df_iphone = ((df['device'].value_counts()['iPhone'])/len(df))*100
pct_df_android = ((df['device'].value_counts()['Android'])/len(df))*100

print(pct_df_iphone)
print(pct_df_android)


64.48429895326355
35.515701046736446


The percentage of missing values by each device is consistent with their representation in the data overall.

There is nothing to suggest a non-random cause of the missing data.

Examine the counts and percentages of users who churned vs. those who were retained. How many of each group are represented in the data?

In [11]:
# Calculate counts of churned vs. retained
c_churned = df['label'].value_counts()['churned']
c_retained = df['label'].value_counts()['retained']

print('churned = ') 
print(c_churned)

print('retained = ') 
print(c_retained)



churned = 
2536
retained = 
11763


This dataset contains 82% retained users and 18% churned users.

Next, compare the medians of each variable for churned and retained users. The reason for calculating the median and not the mean is that you don't want outliers to unduly affect the portrayal of a typical user. Notice, for example, that the maximum value in the `driven_km_drives` column is 21,183 km. That's more than half the circumference of the earth!

In [12]:
# Calculate median values of numeric columns for churned and retained users
med_label = not_null_values.groupby(by='label').median(numeric_only=True)

print(med_label)


              ID  sessions  drives  total_sessions  n_days_after_onboarding  \
label                                                                         
churned   7477.5      59.0    50.0      164.339042                   1321.0   
retained  7509.0      56.0    47.0      157.586756                   1843.0   

          total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
label                                                                        
churned                     84.5                    11.0       3652.655666   
retained                    68.0                     9.0       3464.684614   

          duration_minutes_drives  activity_days  driving_days  
label                                                           
churned               1607.183785            8.0           6.0  
retained              1458.046141           17.0          14.0  


This offers an interesting snapshot of the two groups, churned vs. retained:

Users who churned averaged ~3 more drives in the last month than retained users, but retained users used the app on over twice as many days as churned users in the same time period.

The median churned user drove ~200 more kilometers and 2.5 more hours during the last month than the median retained user.

It seems that churned users had more drives in fewer days, and their trips were farther and longer in duration. Perhaps this is suggestive of a user profile. Continue exploring!

Calculate the median kilometers per drive in the last month for both retained and churned users.

Begin by dividing the `driven_km_drives` column by the `drives` column. Then, group the results by churned/retained and calculate the median km/drive of each group.

In [13]:
# Add a column to df called `km_per_drive`
df['km_per_drive'] = df['driven_km_drives']/df['drives']

# Group by `label`, calculate the median (numeric_only=True), and isolate for km per drive
median_km_per_drive = df.groupby(by='label').median(numeric_only=True)[['km_per_drive']]

median_km_per_drive.head()

km_per_drive
label                 
churned      74.109416
retained     75.014702

The median retained user drove about one more kilometer per drive than the median churned user. How many kilometers per driving day was this?

To calculate this statistic, repeat the steps above using `driving_days` instead of `drives`.

In [14]:
# Add a column to df called `km_per_driving_day`
df['km_per_driving_day'] = df['driven_km_drives']/df['driving_days']

# Group by `label`, calculate the median (numeric_only=True), and isolate for km per driving day
median_km_per_drive_days = df.groupby(by='label').median(numeric_only=True)[['km_per_driving_day']]
                                        
median_km_per_drive_days.head()

km_per_driving_day
label                       
churned           697.541999
retained          289.549333

Now, calculate the median number of drives per driving day for each group.

In [15]:
# Add a column to df called `drives_per_driving_day`
df['drives_per_driving_day'] = df['drives']/df['driving_days']

# Group by `label`, calculate the median (numeric_only=True), and isolate for drives per driving day
median_drives_per_driving_day = df.groupby(by='label').median(numeric_only=True)[['drives_per_driving_day']]

median_drives_per_driving_day.head()

drives_per_driving_day
label                           
churned                  10.0000
retained                  4.0625

The median user who churned drove 698 kilometers each day they drove last month, which is almost ~240% the per-drive-day distance of retained users. The median churned user had a similarly disproporionate number of drives per drive day compared to retained users.

It is clear from these figures that, regardless of whether a user churned or not, the users represented in this data are serious drivers! It would probably be safe to assume that this data does not represent typical drivers at large. Perhaps the data&mdash;and in particular the sample of churned users&mdash;contains a high proportion of long-haul truckers.

In consideration of how much these users drive, it would be worthwhile to recommend to Waze that they gather more data on these super-drivers. It's possible that the reason for their driving so much is also the reason why the Waze app does not meet their specific set of needs, which may differ from the needs of a more typical driver, such as a commuter.

Finally, examine whether there is an imbalance in how many users churned by device type.

Begin by getting the overall counts of each device type for each group, churned and retained.

In [16]:
# For each label, calculate the number of Android users and iPhone users
device_by_label = not_null_values.groupby(by='label')

device_by_label['device'].value_counts()


label     device 
churned   iPhone     1645
          Android     891
retained  iPhone     7580
          Android    4183
Name: count, dtype: int64

Now, within each group, churned and retained, calculate what percent was Android and what percent was iPhone.

In [17]:
# For each label, calculate the percentage of Android users and iPhone users

a = not_null_values[(not_null_values['label'] == 'churned') & (not_null_values['device'] == 'iPhone')]

b = not_null_values[(not_null_values['label'] == 'churned') & (not_null_values['device'] == 'Android')]

c = not_null_values[(not_null_values['label'] == 'retained') & (not_null_values['device'] == 'iPhone')]

d = not_null_values[(not_null_values['label'] == 'retained') & (not_null_values['device'] == 'Android')]


c_churned = not_null_values['label'].value_counts()['churned']
c_retained = not_null_values['label'].value_counts()['retained']

churn_iphone = ((len(a))/(c_churned))*100
churn_android = ((len(b))/(c_churned))*100
retained_iphone = ((len(c))/(c_retained))*100
retained_android = ((len(d))/(c_retained))*100

print(churn_iphone)
print(churn_android)
print(retained_iphone)
print(retained_android)


64.86593059936908
35.13406940063091
64.4393437048372
35.5606562951628


In [18]:
# Answer to question 6

x = ((len(a))/((len(a)+len(c))))*100

y = ((len(b))/((len(b)+len(d))))*100

print (x,y)

17.831978319783197 17.560110366574694


The ratio of iPhone users and Android users is consistent between the churned group and the retained group, and those ratios are both consistent with the ratio found in the overall dataset.

<img src="images/Construct.png" width="100" height="100" align=left>

## **PACE: Construct**

**Note**: The Construct stage does not apply to this workflow. The PACE framework can be adapted to fit the specific requirements of any project.



<img src="images/Execute.png" width="100" height="100" align=left>

## **PACE: Execute**

Consider the questions in your PACE Strategy Document and those below to craft your response:

### **Task 3. Conclusion**

Recall that your supervisor, May Santer, asked you to share your findings with the data team in an executive summary. Consider the following questions as you prepare to write your summary. Think about key points you may want to share with the team, and what information is most relevant to the user churn project.

**Questions:**

1. Did the data contain any missing values? How many, and which variables were affected? Was there a pattern to the missing data?

2. What is a benefit of using the median value of a sample instead of the mean?

3. Did your investigation give rise to further questions that you would like to explore or ask the Waze team about?

4. What percentage of the users in the dataset were Android users and what percentage were iPhone users?

5. What were some distinguishing characteristics of users who churned vs. users who were retained?

6. Was there an appreciable difference in churn rate between iPhone users vs. Android users?





1. There were a total of 700 rows that contained missing values for the ‘label’ variable. There was no major difference between the rows that had missing values in comparison to the rows that did not.

2. The reason for calculating the median and not the mean is that you don't want outliers to unduly affect the portrayal of a typical user.

3. Collect more data on churned users as they appear to be long-haul truck drivers whose needs are not satisfied by the Waze app.

4. 64.5% of all users had iPhones, 35.5% of all users had Android devices

5. There were 4.6 times more retained users than churned users, retained users used the app on over twice as many days as churned users in the same time period, the median churned user drove ~200 more kilometers and 2.5 more hours during the last month than the median retained user, churned users had more drives in fewer days, and their trips were farther and longer in duration

6. Out of all churned users, 64.95 were iPhone users and 35.1% were Android users. 


**Congratulations!** You've completed this lab. However, you may not notice a green check mark next to this item on Coursera's platform. Please continue your progress regardless of the check mark. Just click on the "save" icon at the top of this notebook to ensure your work has been logged.